### PR TITLE
Merging the code paths for interactive and non-interactive co/fixpoints

### PR DESCRIPTION
**Pull request by herbelin**
This is the first part of #18811. A priori ready to be merged.

### CODE REVIEW

**Comment by herbelin** [comment]
@coqbot run full ci

**Comment by Alizter** [comment]
I haven't read the code in full detail, but is it really a good idea to use option types here? I understand that you are using them for flow control in case there are errors creating the fixpoints. But in these cases, wouldn't it be more useful to instead return some result type with say a specific error variant so that we can give users a good error message later rather than asserting false?

On the other hand, if it ought to be impossible to assert false in these situations then shouldn't the types reflect this impossibility?

**Comment by herbelin** [comment]
> is it really a good idea to use option types here

This is not for errors, but for returning an optional proof state, but, indeed, sharing the two possible results  requires an `assert`. My idea is that the use of the assert is local enough (it is `opens`, we pass `pm:None` and should find back `Some` and `None`, it is not `opens, we pass `pm:Some` and we should find back `None` and `Some`) so that it is clear that it is related. (The problem is somehow similar to situations like `match List.map f [a] with [b] -> ... | _ -> assert false`.)

It would probably be possible to avoid the `assert` and `Option.get` by using dependent types. I don't know if I have the courage to do so, but why not on the principle.

File: vernac/comFixpoint.ml [hunk 1]
```diff
@@ -267,25 +267,24 @@ let build_recthms ~indexes fixnames fixtypes fiximps =
...
-    Declare.declare_mutual_definitions ~cinfo ~info ~opaque:false ~uctx
-      ~possible_guard ~bodies:(fixdefs,fixrs) ?using ()
-  in
-  ()
+let declare_recursive ?indexes ?scope ?clearbody ~poly ?typing_flags ?user_warns ?using ((fixnames,fixrs,fixdefs,fixtypes),udecl,uctx,fiximps) ntns =
+  let fix_kind, possible_guard, cinfo = build_recthms ~indexes fixnames fixtypes fiximps in
+  let kind = Decls.IsDefinition fix_kind in
+  let info = Declare.Info.make ?scope ?clearbody ~kind ~poly ~udecl ?typing_flags ?user_warns ~ntns () in
+  try
+    let fixdefs = List.map Option.get fixdefs in
```
Comment (ppedrot) [inline vernac/comFixpoint.ml:275]: Do not rely on Option.get raising exceptions, instead use Option.List.map.

**Comment by herbelin** [comment]
@coqbot run full ci

**Comment by herbelin** [comment]
Assuming @Alizter is ok, it will help further work if this is merged (the issue of having to pass a `pm` or not, or deciding statically if it will produce a proof or not is somehow a different question).

**Comment by Alizter** [comment]
I'm fine with the direction since it will be refactored further. Hopefully when the refactoring is done it will lead to some more simplification in the implementation.

**Comment by herbelin** [comment]
@coqbot run full ci

**Review by SkySkimmer — APPROVED** [review]

**Comment by SkySkimmer** [comment]
@coqbot merge now

**Pull request merged**